File: app/src/androidTest/java/com/vgrec/espressoexamples/SpinnerSelectionTest.java
package com.vgrec.espressoexamples;

import com.vgrec.espressoexamples.bases.TestBase;

import org.junit.Before;
import org.junit.Test;

import static androidx.test.espresso.Espresso.onData;
import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static com.vgrec.espressoexamples.matchers.CustomMatchers.withAdaptedData;
import static com.vgrec.espressoexamples.matchers.CustomMatchers.withItemContent;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.Matchers.allOf;
import static org.hamcrest.Matchers.instanceOf;
import static org.hamcrest.Matchers.not;


/**
 * @author  HDunn, Modifed on 4/30/21.
 */

public class SpinnerSelectionTest extends TestBase {

    public static final String INVALID_COUNTRY_NAME = "NoSuchCountry";
    public static final String VALID_COUNTRY_NAME = "Moldova";
    public static final String FIRST_ITEM_TEXT = "Select your country";

    @Before
    public void navigateToSpinnerSelection(){
        // From Main navigate to spinner selection
        onView(withId(R.id.spinner_selection_button)).perform(click());
    }

    @Test
    public void testCountryNotInList() {
        onView(withId(R.id.countries_spinner)).check(matches(not(withAdaptedData(withItemContent(INVALID_COUNTRY_NAME)))));
    }

    @Test
    public void testLabelDoesNotChangeIfFirstItemSelected() {
        // Click on the Spinner
        onView(withId(R.id.countries_spinner)).perform(click());

        // Click on the first item from the list, which is a marker string: "Select your country"
        onData(allOf(is(instanceOf(String.class)))).atPosition(0).perform(click());

        // Check that the country label is not updated.
        onView(withId(R.id.country_label)).check(matches(not(withText(FIRST_ITEM_TEXT))));
    }

    @Test
    public void testLabelUpdatesIfValidCountrySelected() {
        // Click on the Spinner
        onView(withId(R.id.countries_spinner)).perform(click());

        // Select a country from the list
        onData(allOf(is(instanceOf(String.class)), is(VALID_COUNTRY_NAME))).perform(click());

        // Check that the country label is updated with selected country
        onView(withId(R.id.country_label)).check(matches(withText(VALID_COUNTRY_NAME)));
    }
}

File: app/src/androidTest/java/com/ui/espresso/SpinnerSelectionTest.kt
package com.ui.espresso

import androidx.test.espresso.Espresso
import androidx.test.espresso.action.ViewActions
import androidx.test.espresso.assertion.ViewAssertions
import androidx.test.espresso.matcher.ViewMatchers
import com.ui.espresso.bases.TestBase
import com.ui.espresso.matchers.CustomMatchers
import com.ui.espresso.screens.MainScreen
import com.ui.espresso.screens.SpinnerScreen
import org.hamcrest.CoreMatchers
import org.hamcrest.Matchers
import org.junit.Before
import org.junit.Test

/**
 * @author  HDunn, Modified on 4/30/21.
 */
class SpinnerSelectionTest : TestBase() {
    @Before
    fun navigateToSpinnerSelection() {
        // From Main navigate to spinner selection
        MainScreen.tapSpinnerButton()
    }

    @Test
    fun testCountryNotInList() {
        SpinnerScreen.verifyCountyNotInList
    }

    @Test
    fun testLabelDoesNotChangeIfFirstItemSelected() {
        // Click on the Spinner
        SpinnerScreen.tapSpinner()

        // Click on the first item from the list, which is a marker string: "Select your country"
        SpinnerScreen.tapFirstItem()

        // Check that the country label is not updated.
        SpinnerScreen.verifySpinnerDefaultText
    }

    @Test
    fun testLabelUpdatesIfValidCountrySelected() {
        // Click on the Spinner
        SpinnerScreen.tapSpinner()

        // Select a country from the list
        SpinnerScreen.tapCountry()

        // Check that the country label is updated with selected country
        SpinnerScreen.verifyCountrySelected
    }

}
